packages/webgpu/src/Canvas.tsx

```diff
@@ -31,45 +44,92 @@ export interface NativeCanvas {
 };
 
 export interface CanvasRef {
   getContext(contextName: "webgpu"): RNCanvasContext | null;
   getNativeSurface: () => NativeCanvas;
+  whenReady: (callback: () => void) => void;
 }
 
-export const Canvas = forwardRef<CanvasRef, ViewProps>((props, ref) => {
-  const [contextId, _] = useState(() => generateContextId());
+interface Size {
+  width: number;
+  height: number;
+}
 
-  useImperativeHandle(ref, () => ({
-    getNativeSurface: () => {
-      WebGPUNativeModule.createSurfaceContext(contextId);
-      return WebGPUContextRegistry[contextId];
-    },
-    getContext(contextName: "webgpu"): RNCanvasContext | null {
-      if (contextName !== "webgpu") {
-        throw new Error(`[WebGPU] Unsupported context: ${contextName}`);
-      }
-      WebGPUNativeModule.createSurfaceContext(contextId);
-      const nativeSurface = WebGPUContextRegistry[contextId];
-      if (!nativeSurface) {
-        return null;
+const useSizeFabric = (ref: RefObject<View>) => {
+  const [size, setSize] = useState<null | Size>(null);
+  useLayoutEffect(() => {
+    if (!ref.current) {
+      throw new Error("Canvas ref is null");
+    }
+    ref.current.measureInWindow((_x, _y, width, height) => {
+      setSize({ width, height });
+    });
+  }, [ref]);
+  return { size, onLayout: undefined };
+};
+
+const useSizePaper = (_ref: RefObject<View>) => {
+  const [size, setSize] = useState<null | Size>(null);
+  const onLayout = useCallback<(event: LayoutChangeEvent) => void>(
+    ({
+      nativeEvent: {
+        layout: { width, height },
+      },
+    }) => {
+      if (size === null) {
+        setSize({ width, height });
       }
-      const ctx = RNWebGPU.MakeWebGPUCanvasContext(nativeSurface);
-      return ctx;
     },
-  }));
-
-  useEffect(() => {
-    return () => {
-      delete WebGPUContextRegistry[contextId];
-    };
-  }, [contextId]);
+    [size],
+  );
+  return { size, onLayout };
+};
 
-  return <WebGPUNativeView {...props} contextId={contextId} />;
-});
+export const Canvas = forwardRef<CanvasRef, ViewProps>(
+  ({ onLayout: _onLayout, ...props }, ref) => {
+    const viewRef = useRef(null);
+    const FABRIC = RNWebGPU.fabric;
+    const useSize = FABRIC ? useSizeFabric : useSizePaper;
+    const [contextId, _] = useState(() => generateContextId());
+    const cb = useRef<() => void>();
+    const { size, onLayout } = useSize(viewRef);
+    useEffect(() => {
+      if (size && cb.current) {
+        cb.current();
+      }
+    }, [size]);
+    useImperativeHandle(ref, () => ({
+      getNativeSurface: () => {
+        return RNWebGPU.getNativeSurface(contextId);
+      },
+      whenReady(callback: () => void) {
+        cb.current = callback;
+      },
+      getCallback: () => {
+        return cb.current;
+      },
+      getContext(contextName: "webgpu"): RNCanvasContext | null {
+        if (contextName !== "webgpu") {
+          throw new Error(`[WebGPU] Unsupported context: ${contextName}`);
+        }
+        if (size === null) {
```

Comment:
here it throws an error if the size is not available yet. @piaskowyk is that what you meant? I should add it to getNativeSurface too